### PR TITLE
Support CUDA Unified Memory

### PR DESCRIPTION
**Pull request by 0x804d8000**
# Motivation

Occasionally we found some overlarge samples (e.g. long sequence during training a Transformer-based model) that could cause OOM. Although carefully filtering them out or applying some manually-designed transformation can reduce OOM possibility, these methods can hardly eliminate the issue completely.

# Resolution here

This PR adds support for [CUDA Unified Memory](https://developer.nvidia.com/blog/unified-memory-cuda-beginners/), or UVM, to PyTorch, such that we effectively makes device RAM "swappable" to system RAM. We usually have plenty of system RAM (e.g., several terabytes) compared to device RAM (tens of gigabytes), therefore UVM brings us a much large safe margin before hitting OOM.

The feature can be optionally enabled by setting `PYTORCH_CUDA_USE_UVM=1` in environment variables.

# Performance implications

UVM degrades CUDA performance, but how much does it affects depends on the actual workload.

From our earlier experience, UVM hardly harms CUDA kernels, however, D2H/H2D memcpy are slowed down to a certain degree. Therefore, for memcpy-heavy workload, UVM can makes things much slower.

However, for training workload with an adequately large batch size, the performance should degrade much. This is also empirically verified by [QLoRA: Efficient Finetuning of Quantized LLMs](https://arxiv.org/abs/2305.14314) ("Paged Optimizers"). This makes UVM appealing to training workload.

# Possibly related

https://github.com/pytorch/pytorch/issues/98481 seems related, which also tried to add UVM support to PyTorch, albeit via a difference approach.

cc @ptrblck

### CODE REVIEW

File: c10/cuda/CUDACachingAllocator.cpp [hunk 1]
```diff
@@ -124,6 +124,11 @@ using stream_set = ska::flat_hash_set<cuda::CUDAStream>;
 
 using StatTypes = std::array<bool, static_cast<size_t>(StatType::NUM_TYPES)>;
 
+bool use_cuda_uvm() {
+  static bool use_uvm = getenv("PYTORCH_CUDA_USE_UVM") != nullptr;
```
Comment (Skylion007) [inline c10/cuda/CUDACachingAllocator.cpp:128]: Wouldn't this still evaluate to true if `PYTORCH_CUDA_USE_UVM=0`?

**Comment by ezyang** [comment]
We're going to have to work out some testing strategy, at least.

**Comment by 0x804d8000** [comment]
Hi @Skylion007 @ezyang, I revised the patch accordingly, would you mind taking a look at it again?

**Comment by ezyang** [comment]
One thing I'm not sure about, overall design wise, is whether or not to do this as a global flag or not. I understand that having it as a global flag is convenient for the use case "I just want to change a single flag and suddenly my program can use more memory without OOMing", but the underlying cudaMallocManaged function can in principle be used in a more fine-grained manner by having a different allocator. This is what the previous approach you linked to does, and it also appears to be how TF implemented it: https://github.com/tensorflow/tensorflow/commit/cd4f5840

Related https://github.com/pytorch/pytorch/issues/44380

If we do go the global flag route, instead of having a completely separate envvar you should unify the handling with PYTORCH_CUDA_ALLOC_CONF see CachingAllocatorConfig in c10/cuda/CUDACachingAllocator.cpp

cc @ngimel

**Comment by 0x804d8000** [comment]
I’m a little concerned with the "new allocator" way. Unlikely CUDA stream ordered allocator (cudaMallocAsync, which we already supported as a separate allocator backend), UVM does not natively recognizes CUDA stream dependencies. This means to work with CUDA streams, we either have to pay the cost of heavy device-wide synchronization implied by cudaFree, or have to implement another "caching" layer for UVM allocator. 

Use PYTORCH_CUDA_ALLOC_CONF to enable UVM sounds reasonable to me, I’ll see if I can get this working.

**Comment by Skylion007** [comment]
Tensorflow enables CUDA UVM by setting a memory limit >=1 so they also have it as a single global flag (for the TF.session at least) @ezyang

**Comment by 0x804d8000** [comment]
Now UVM is enabled by setting `use_uvm=True` in `PYTORCH_CUDA_ALLOC_CONF`. @ezyang

**Comment by ptrblck** [comment]
Without prefetching I would assume to see a poor performance, so do we have any real measurements besides the linked paper?

Also, why doesn't [`torch.cuda.CUDAPluggableAllocator`](https://pytorch.org/docs/stable/generated/torch.cuda.CUDAPluggableAllocator.html#torch.cuda.CUDAPluggableAllocator) work for this use case? 

@syed-ahmed @eqy Could you take a look at it, too, please?

**Comment by 0x804d8000** [comment]
Unless we’re over subscribing device memory, prefetching is not relevant here. Because we always explicitly copy tensors to device via cudaMemcpy before using them, tensors are always resident on device memory regardless of whether UVM is used. 

As I said in the “Performance implications” section, CUDA kernels don’t see a slow down with UVM, assuming we manually copied data before using them (which is the case in PyTorch). The difference shows up  when doing H2D/D2H copy, these copies do see a performance hit. I’m not sure why it’s the case, maybe UVM needs more interaction with the driver or something. 

So the key here is to avoid frequent D2H/H2D copy. In the training scenario, we naturally not “cudaMemcpy API bound”, so things should be fine. Personally I tested the patch with a non-standard GPT model, with B=1 I saw a 2x slow down, however, once I increased B to 32, it reaches 96% throughout of non-UVM version.

**Comment by 0x804d8000** [comment]
The issue with [torch.cuda.CUDAPluggableAllocator](https://pytorch.org/docs/stable/generated/torch.cuda.CUDAPluggableAllocator.html#torch.cuda.CUDAPluggableAllocator) is that it’s not currently as full featured as the “native” allocator, see #98481. From my experience I also saw some issue such as torch.cuda.memory_allocated() doesn’t work. I’m ok with the pluggable allocator way if it works seamlessly as the native one.

**Comment by ezyang** [comment]
If people like the global flag, I guess I can deal with the global flag. I'd like it if NVIDIA peeps could verify the rest of the implementation.

**Comment by syed-ahmed** [comment]
https://github.com/pytorch/pytorch/issues/98481 raises a good concern and I think what it's saying is CUDAPluggableAllocator needs a little more work such that external allocators can be a drop-in replacement for the built-in allocators in PyTorch. To elaborate a little,
- When `torch.backends.cudnn.benchmark = True`, cudnn backend will hit the benchmark path that uses `cacheInfo`. Since only alloc and free are provided with the pluggable allocator, cacheInfo call raises exception (in pluggable allocator) when cudnn benchmark is used, rightfully so - cacheInfo's job is to provide a reasonable estimate for cudnn workspace size and different allocators can implement their own logic for cacheInfo (e.g. CudaMallocAsyncAllocator's cacheInfo is different than CudaCachingAllocator's). We could add a way for the user to provide the cacheInfo function, just like they do for alloc and free (may be also refactor the cacheInfo function because a pluggable allocator can be a non-caching allocator).
- Functions related to memory statistics such as torch.cuda.memory_allocated(), memory_snapshot etc. are also allocator specific. 

So even though CUDAPluggableAllocator is a good use case for this PR, it's understandable to enable UVM with the environment variable.

Questions for @0x804d8000 : 
- how should users interpret memory_stats / memory_allocated with your PR when `use_uvm` is true? Can you give some examples? Do I understand correctly that with your PR, cudaMallocManaged will "spill" the memory to cpu memory  when gpu memory is exhausted? 
- Is there any change in the chaching mechanism of the CUDACachingAllocator when using uvm? I see you have `cudaMallocManaged` in `cudaMallocMaybeCapturing` which is taking the cached path. When memory is oversubscribed and part of the memory is on cpu, is that covered in the gpu cached segment? I'm not sure about the implications of that yet, but it sounds wrong if the CUDACachingAllocator is supposed to have cached allocs that are for gpu memory, but now with uvm it's both cpu and gpu? May be @zdevito can comment.
- Can you comment on this issue about using cudaMallocManaged (from the [fbgemm_gpu repo](https://github.com/pytorch/FBGEMM/blob/eb64159d9206c2a57c43185b53fa02c71c8ca4f6/fbgemm_gpu/src/cumem_utils.cu#L25-L33) ):
   ```
   // Freeing host/uvm memory with cudaFree[Host] requires a cuda context.
   // If a uvm tensor is released from an arbitrary thread without a context
   // then cuda helpfully create a new default context on the default device.
   // If we have not used the default device before in this process cuda
   // needs to also allocate a device context. However creating a device
   // context requires device resources and may fail with out of memory error
   // causing  cudaFree[Host] to fail with out of memory error.
   // The solution is simply to remember the device from the allocation context
   // and set the correct device in the thread before calling cudaFree[Host]
   ```
   Does this happen with your PR? @ezyang is this a concern for this PR?
- `cudaMallocManaged` allocates using page faults. Can we see some benchmarks/testing that sheds some light on if there is an overhead? For instance, you can use nsight profile with the options `--cuda-um-cpu-page-faults` and `--cuda-um-gpu-page-faults`.

**Comment by 0x804d8000** [comment]
- a) Now `memory_allocated` may get bigger than `device_props.total_memory`, when over-subscription occurs. For people who use UVM, I think they should already be expecting this, as the only reason they enable UVM is for over-subscription. The rest fields may be interpreted as if no UVM is involved. b) Sort of. Which memory blocks are "spilled" are dynamically determined, in a pseudo-LRU way. Memory may get migrated back-and-forth between device memory and system memory. You may think device memory as a "huge cache" for system memory when UVM is used.
- I think the most benefit of caching allocator is avoiding the heavyweight `cudaFree`, which incurs a device-wide synchronization. Regardless of the allocation mechanism, I think we always want such benefit. So yes, I think caching allocator should be used here, not merely for "caching some device memory", but for "caching memory allocated from cudart so that we don't have to pay the cost of `cudaFree` on each deallocation". Another practical reason, or rather, limitation, that we want to treat all memory allocated via UVM the same way (being cached) is that we cannot tell which part of memory are on device and which are on system memory, as the placement is dynamically determined, therefore we can hardly treat them separately.
- Honestly I don't have much to comment on this issue, the reasons are: 1) I have no experience with multithreaded training. If multiple devices are used, we run multiple instances of PyTorch (i.e., multiple python processes) via `torchrun`, assigning each instance a device. 2) With UVM enabled, device memory allocated by PyTorch are all "migrateble" and can be migrated to host memory were there such a need, so we have much greater headroom for things that needs "pinned device memory", including cuda context initialization, cudnn & cublas workspace, ... . Considering that usually we have tens of gigabytes allocated via PyTorch and all of them can be migrated away for make room for cuda context initialization, it's highly unlikely we can still run out of device memory and fail the initialization.
- It relies on page faults only if we do not explicit copy memory between device and host. This is not case with PyTorch. We already have necessary `cudaMemcpyAsync` in place, so effectively all page faults / prefetching / ... are "done manually", so page faults is not relevant here, just as how prefetching is not relevant. This can be empirically verified with the snippet below.

```cuda
#include <cuda_runtime.h>
#include <stdio.h>
#include <vector>

__global__ void foo(int *ptr, int n) {
  int x = threadIdx.x + blockIdx.x * blockDim.x;
  if (x < n) {
    ptr[x] = 1;
  }
}

int main(int argc, char **argv) {
  int *ptr;
  int N = 100'000'000;
  printf("%d\n", cudaMallocManaged((void **)&ptr, N * sizeof(int)));
  if (argc > 1) {
    std::vector<int> v(N);
    printf("%d\n",
           cudaMemcpy(ptr, v.data(), N * sizeof(int), cudaMemcpyHostToDevice));
  }
  foo<<<(N + 1023) / 1024, 1024>>>(ptr, N);
  printf("%d\n", cudaDeviceSynchronize());
}
```

If no extra argument is provide, we indeed incur page faults:

image

However, if extra argument is provided (and invokes `cudaMemcpy`) when running the program, we see no page fault in nsys' result:

image

**Comment by syed-ahmed** [comment]
> Now memory_allocated may get bigger than device_props.total_memory

Can you please show using a pytorch snippet and the output you see with your PR?

> b) Sort of. Which memory blocks are "spilled" are dynamically determined, in a pseudo-LRU way. Memory may get migrated back-and-forth between device memory and system memory.

Can you elaborate what you mean by dynamically determined and pseudo-LRU way? Memory migrating back-and-forth is the performance downside of cudaMallocManaged that we want to avoid, and that's why cudaMemAdvise and cudaMemPrefetchAsync are there, so that users can provide migration policies. Also, what happens in a backward call?

> we want to treat all memory allocated via UVM the same way (being cached) is that we cannot tell which part of memory are on device and which are on system memory, as the placement is dynamically determined, therefore we can hardly treat them separately.

This is what I'm concerned with the design. If a user says allocate my tensor on `device="cuda"` and pytorch has migrated part of that memory to "cpu" by using cudaMallocManaged, aren't we breaking the user's intent? From @ezyang's blog, "the distinguishing characteristic of a device is that it has its own allocator, that doesn't work with any other device".

> It relies on page faults only if we do not explicit copy memory between device and host. 

Yes. I think the point here is it page faults. Lets say you have 24 GB device memory and you need a 30 GB tensor, when you allocate a 30 GB tensor with cudaMallocManaged and access it in your kernel, won't you page fault? That's why it's a little surprising that there is no usage of cudaMemAdvise or cudaMemPrefetchAsync in this PR. 

For the attached paper, I see they indeed use cudaMemPrefetchAsync:
https://github.com/TimDettmers/bitsandbytes/blob/18e827d666fa2b70a12d539ccedc17aa51b2c97c/bitsandbytes/optim/optimizer.py#L314-L322. So I think we need to see more results from your PR.

For the GPT model, was it an inference run or fine tuning?

**Comment by 0x804d8000** [comment]
I think I see your point now. You're concerned about the performance when the user does over-subscribe VRAM, right?

Yes, they will suffer.

Let me reiterate the motivation of this PR:

> Occasionally we found some overlarge samples (e.g. long sequence during training a Transformer-based model) that could cause OOM. Although carefully filtering them out or applying some manually-designed transformation can reduce OOM possibility, these methods can hardly eliminate the issue completely.

So UVM is not intended as a solution for constantly VRAM over-subscription, but an once-and-for-all mitigation for occasional OOMs caused by overlarge sequence or alike. Upon seeing such inputs, the expected behavior changes from:

- OOM occurs and the training process crashes, requiring manually intervening.

to:

- Over-subscription occurs and the training speed of **this** batch slows down, due to UM page faults, but no interruption to the training process would occur, and the rest (batches before and batches after) trains at the near-optimal (no-UM) speed [1].

In case the training workload is constantly larger than VRAM, instead of manually prefetching, CPU offload may be a better choice. Either way we'd be limited by PCIe bandwidth, and with explicitly copying (offloading) between device and host, we should get the optimal result.

To conclude, the PR is a once-and-for-all mitigation for occasional OOMs, not a solution for workloads that are constantly larger than VRAM.

Hopefully this clarifies something for your concern #2 & #4.

As for #3:

No it's not PyTorch who migrate the UVM memory, it's the CUDA driver. PyTorch has no control on the placement policy (besides `cudaMemAdvise` or alike). For user intent, we still allocate the memory in the CUDA way, and the data do resides on CUDA driver (except for occasional over-subscription). Also, the contract (I mean, interface to the user) that "that doesn't work with any other device" should still hold. Accessing a CUDA tensor, even if it's allocated with UVM and "physically" accessible, without first migrating (in PyTorch's sense this time, i.e., `tensor.cpu()`) it to CPU, is still a violation of our design. [2]

The snippet asked:

```python
Python 3.9.2 (default, Feb 28 2021, 17:03:44) 
[GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> print(torch.cuda.get_device_properties(0).total_memory)
85198045184
>>> a = torch.empty(85198045184 * 2, dtype=torch.int8, device='cuda')
>>> print(torch.cuda.memory_allocated())
170396090368
>>> 
```

---

1: Actually we may even get better training speed than non-UM case, as, as a side effect, UVM "mitigates" memory fragmentation issue. Those "fragmented but not used" memory blocks will likely be eventually migrated to system memory by CUDA driver, leaving more VRAM for us to use and allowing a larger batch size.

2: Well we may some day need to revise this design as vendors are trying to "unify" the memory space, such as NVIDIA's Grace Hopper Superchip, AMD's MI300, ... . But this is not relevant to this PR.

**Comment by 0x804d8000** [comment]
Oh I see why you said "pytorch has migrated part of that memory". Did you mean explicitly migrating the memory via `cudaMemAdvise`? We don't do that in this PR. So migration is totally up to the CUDA driver. 

Normally we want all data reside on device memory. Migrating part of data to cpu should only happen when some exceptionally large input is somehow given by the data loader.

Yes I'm aware that we can do some fancy batching strategy so that the input sizes are under a certain limit, but devising a perfect strategy (100% no OOM) is much harder than a good enough one (99% of the time no over subscription). Besides, a perfect safe strategy usually requires such a strict batch size rule that often overestimates VRAM actually required, leading to an unnecessary small batch size, and slows down the overall training speed. With this PR, we only need a good enough strategy, leaving the occasional OOM for UVM to handle, and only pay the cost of being slow of that single batch.

**Comment by 0x804d8000** [comment]
@albanD Thanks for the reminder. 

Yes I’m aware of that behavior, and I’ve been “blessed” by it. After allocating terabytes of memory, a CUDA kernel miserably raised an “illegal memory access” error…

This is now mitigated by the addition of c10/cuda/CUDACachingAllocator.cpp:1385. When UVM is enabled, the allocator now proactively reclaims memory if there has been too much of memory being reserved.

**Comment by zdevito** [comment]
Turning on unified memory increases the complexity of the mental model of performance for running networks. Previously the performance of a network was uncoupled to how much memory has been allocated.  With unified memory, OOMs are disabled entirely but slowdowns will be introduced that get worse with more memory. I can imagine having situations where a developer increases memory inadvertently and is left with a hard-to-debug slowdown. Unified memory is not immediately compatible with expandable_segments, which is also trying to enable more complete use of memory at the limits, so I can see places where it will not be clear how to choose which approach to use.

Until we see large adoption of unified memory to solve memory problems, this allocator seems more appropriate as a custom allocator where it is very clear that non-standard behavior has been enabled. 

So I think the right path forward is to fix the issues that make the pluggable allocator unsuitable for this case. In particular, it makes sense for the pluggable allocator to fallback to the existing allocator for functionality it doesn't implement, and for the the pluggable allocator to be able to override how non-cached memory is obtained from cuda even when falling back to the existing caching approach. For unified memory, the custom allocator would then use the standard implementations of cached_alloc/cached_free, but override uncached_alloc/uncached_free to back that memory with managed memory.

**Comment by Allen050329** [comment]
Hi, uh, I know I'm a bit late to the party, but the actual reasons I suggested using the pluggable allocator were: 1. so other allocators that pop up in the future could be used next to seamlessly, and 2. I thought it'd require less code, which seems like a misconception on my part
@0x804d8000 Thanks for the link, I'm checking it out right now

**Comment by rmast** [comment]
> So I guess the preferred way is to revising the pluggable allocator, closing the PR then.
> 
> For anyone who is interested in the patch, I'll keep it in [my fork's uvm branch](https://github.com/0x804d8000/pytorch/tree/uvm).
> 
> Thank you all for the valuable comments. 🙏

Thanks man! I can finally EasyOCR an 300 dpi A4 page on an NVIDIA GT 1030 with only 2 GB within a reasonable time on an AMD FX6300 without AVX2!

**Pull request closed without merging**